### DISCUSSION TITLE
micropython.mk being ignored

### DISCUSSION THREAD

**Discussion by ma261065** [micropython/micropython#12700] [ESP32] [answered]
Hi all,

I'm building a C module, and have set up the folder structure as shown at https://docs.micropython.org/en/latest/develop/cmodules.html

`my_project/`
`└── modules/`
`    ├── example1/`
`    │   ├── example1.c`
`    │   ├── micropython.mk`
`    │   └── micropython.cmake`
`    └── micropython.cmake`

My module compiles fine, and I can successfully call it from MicroPython code.

I wanted to add some compiler flags, so put them in `micropython.mk`, but they didn't seem to take effect.

After further troubleshooting, it seems that the whole `micropython.mk` file is being ignored. I can put any garbage in there, and it will still compile successfully. I can even rename the file, and my module will still compile. I am doing a clean between compiles.

What am I missing to make it use the `micropython.mk` file as per the documentation?

**Comment by andrewleech** [comment, marked as answer]
Which port / chip are you compiling for? rp2 and esp use cmake so extra definitions need to go in the `micropython.cmake` fragment. Other make based ports use `micropython.mk`

**Reply by ma261065** [reply]
It's an ESP32.

Thanks, I didn't realise that different ports behaved differently. So if I want to add compiler flags, I can just use the same syntax in `micropython.cmake`?

e.g. `CFLAGS_USERMOD += -O3`

**Reply by andrewleech** [reply]
I'm not sure what the syntax is for the cmake fragments sorry, it's usually quite different to makefile ones.

**Reply by ma261065** [reply]
You're right.

From researching more, it seems that the syntax is as follows:

`# Set compiler options for specific source files`
`set_source_files_properties(${CMAKE_CURRENT_LIST_DIR}/example1.c PROPERTIES COMPILE_OPTIONS -O3)`

Thanks for the help in solving the mystery!

**Reply by ma261065** [reply]
Well, for the life of me I can't get any compiler flags set in `micropython.cmake` to actually work.

For example, if I set various -O flags using `set_source_files_properties` (e.g. -O3 or -Os) the compiled code size is exactly the same. I can put invalid options in there (e.g. -O9) and the compiler doesn't complain.

However, if I put `#pragma GCC optimize "O3"` or `#pragma GCC optimize "Os"` in the source file I can see the code size change.

How do I get flags in `micropython.cmake` to work?

**Reply by andrewleech** [reply]
Yeah I hate cmake. For something _so powerful_ it's got too many obscurities in how to make things actually work.

Can you share your `micropython.cmake` contents with your attempts to set the flag?

**Reply by ma261065** [reply]
Here it is. 

I tried `COMPILE_OPTIONS` as well as `COMPILE_FLAGS`, with and without quotes around the flag, with and without the leading `-` on the option.

`# Create an INTERFACE library for our C module.`
`add_library(usermod_Decoder INTERFACE)`

`# Add our source files to the lib`
`target_sources(usermod_Decoder INTERFACE`
`    ${CMAKE_CURRENT_LIST_DIR}/example1.c`
`    ${CMAKE_CURRENT_LIST_DIR}/example2.cpp`
`)`

`# Add the current directory as an include directory.`
`target_include_directories(usermod_Decoder INTERFACE`
`    ${CMAKE_CURRENT_LIST_DIR}`
`)`

`# Set compiler options for specific source files`
`set_source_files_properties(${CMAKE_CURRENT_LIST_DIR}/example1.c PROPERTIES COMPILE_OPTIONS -O3)`
`set_source_files_properties(${CMAKE_CURRENT_LIST_DIR}/example2.cpp PROPERTIES COMPILE_OPTIONS -Os)`

`# Link our INTERFACE library to the usermod target.`
`target_link_libraries(usermod INTERFACE usermod_Decoder)`

**Reply by andrewleech** [reply]
What you've done certainly _looks_ about right... I'm seeing a couple of references to `COMPILE_FLAGS` instead of `COMPILE_OPTIONS`, maybe try
```
set_source_files_properties(${CMAKE_CURRENT_LIST_DIR}/example1.c PROPERTIES COMPILE_FLAGS -O3)
```

However it also looks like built-in settings might get in the way of that, or override that, eg. https://stackoverflow.com/questions/47926404/cmake-per-file-optimizations

**Reply by andrewleech** [reply]
This gives another suggestion: https://jeremimucha.com/2021/02/cmake-fundamentals-part5/ in section "Project configuration interface targets
```
add_library(usermod_Decoder INTERFACE)
target_compile_options(usermod_Decoder INTERFACE -O3)
```
though that'll affect all files in the user c module. Perhaps the "generator" in the example on that page that sets something on just c++ files could be adjusted to only set something on a file with particular name.
